1854/admin command support improvement

laverya opened the pull request.

laverya committed:
Expand validation of admin commands, and accept a greater variety of admin command schemas
added 'Not' predicate that does exactly what you'd think

laverya committed:
Split up verificiation of admin command completeness to improve error messages

laverya committed:
Updated `not` predicate to not return undefined path if setting matched to true
Added test to make sure that some sort of container identification (replicated, swarm, or kubernetes in various styles) is present in admin_commands

dexhorthy commented on src/predicates.ts:
```diff
@@ -1182,3 +1195,23 @@ export class IsNotKubernetesQuantity implements Predicate<any> {
+  public test(root: any): RuleMatchedAt {
+    const result = this.pred.test(root);
+    if (!result.matched && _.isUndefined(result.paths)) {
+      return { matched: !result.matched, paths: [""] };
```
I think having an empty path here might cause unexpected things to happen. What about casting an empty array with `as`?

dexhorthy replied:
or maybe just omit `paths` entirely?

dexhorthy commented on src/predicates.ts:
```diff
@@ -1182,3 +1195,23 @@ export class IsNotKubernetesQuantity implements Predicate<any> {
   }
 }
+
+export class Not<T_El> implements Predicate<T_El> {
```
Nice addition!

laverya committed:
Reduce the possibility of issues caused by predicates asserting matches at empty strings
`tokenize` now drops a single trailing `.` character, if present, before tokenizing a string to avoid including empty tokens

laverya committed:
Add to changelog, bump version
